prefect_kubernetes/credentials.py

```diff
@@ -0,0 +1,133 @@
+"""Module for defining Kubernetes credential handling and client generation."""
+
+from contextlib import contextmanager
+from typing import Generator, Optional
+
+from kubernetes import config as kube_config
+from kubernetes.client import ApiClient, AppsV1Api, BatchV1Api, CoreV1Api
+from kubernetes.config.config_exception import ConfigException
+from prefect.blocks.core import Block
+from prefect.blocks.kubernetes import KubernetesClusterConfig
+
+
+class KubernetesCredentials(Block):
+    """Credentials block for generating configured Kubernetes API clients.
+
+    Attributes:
+        cluster_config: A `KubernetesClusterConfig` block holding a JSON kube
+            config for a specific kubernetes context.
+
+    Examples:
+        Load stored Kubernetes credentials:
+        ```python
+        from prefect_kubernetes.credentials import KubernetesCredentials
+
+        kubernetes_credentials = KubernetesCredentials.load("my-k8s-credentials")
+        ```
+
+        Create resource-specific API clients from KubernetesCredentials:
+        ```python
+        from prefect_kubernetes import KubernetesCredentials
+
+        kubernetes_credentials = KubernetesCredentials.load("my-k8s-credentials")
+
+        kubernetes_app_v1_client = kubernetes_credentials.get_app_client()
+        kubernetes_batch_v1_client = kubernetes_credentials.get_batch_client()
+        kubernetes_core_v1_client = kubernetes_credentials.get_core_client()
+        ```
+
+        Create a namespaced job:
+        ```python
+        from prefect import flow
+        from prefect_kubernetes import KubernetesCredentials
+        from prefect_kubernetes.job import create_namespaced_job
+
+        from kubernetes.client.models import V1Job
+
+        kubernetes_credentials = KubernetesCredentials.load("my-k8s-credentials")
+
+        @flow
+        def kubernetes_orchestrator():
+            create_namespaced_job(
+                body=V1Job(**{"Marvin": "42"}),
+                kubernetes_credentials=kubernetes_credentials
+            )
+        ```
+    """
+
+    _block_type_name = "Kubernetes Credentials"
+    _logo_url = "https://images.ctfassets.net/zscdif0zqppk/oYuHjIbc26oilfQSEMjRv/a61f5f6ef406eead2df5231835b4c4c2/logo.png?h=250"  # noqa
+
+    cluster_config: Optional[KubernetesClusterConfig] = None
+
+    @contextmanager
+    def get_app_client(self) -> Generator[AppsV1Api, None, None]:
+        """Convenience method for retrieving a kubernetes api client for deployment resources
+
+        Returns:
+            Kubernetes api client generator to interact with "deployment" resources.
+        """
+        generic_client = self.get_generic_client()
+        try:
+            yield AppsV1Api(api_client=generic_client)
+        finally:
+            generic_client.rest_client.pool_manager.clear()
```

Comment:
These should be context managers I think.
```suggestion
        with self.get_generic_client() as generic_client:
                try:
                    yield AppsV1Api(api_client=generic_client)
                finally:
                    generic_client.rest_client.pool_manager.clear()
```